table/block_based/block_based_table_builder.cc

```diff
@@ -1518,7 +1519,31 @@ void BlockBasedTableBuilder::WriteMaybeCompressedBlock(
   }
   // Old, misleading name of this function: WriteRawBlock
   StopWatch sw(r->ioptions.clock, r->ioptions.stats, WRITE_RAW_BLOCK_MICROS);
-  const uint64_t offset = r->get_offset();
+
+  uint64_t offset = r->get_offset();
+  // try to align the page to the super alignment size, if enabled
+  if (r->table_options.super_block_align && is_data_block) {
+    auto super_block_alignment_mask =
+        r->table_options.super_block_alignment_size - 1;
+    if ((offset & (~super_block_alignment_mask)) !=
+        ((offset + block_contents.size()) & (~super_block_alignment_mask))) {
+      // new block would cross the super block boundary
+      auto pad_bytes = r->table_options.super_block_alignment_size -
+                       (offset & super_block_alignment_mask);
+      if (pad_bytes <=
+          r->table_options.super_block_alignment_max_padding_size) {
+        io_s = r->file->Pad(io_options, pad_bytes);
+        if (!io_s.ok()) {
+          r->SetIOStatus(io_s);
+          return;
+        }
+        r->pre_compression_size += pad_bytes;
+        offset += pad_bytes;
+        r->set_offset(offset);
```

Comment:
Since I proposed this, I realized a problem: most index block entries assume the data blocks are immediately adjacent to each other. https://github.com/facebook/rocksdb/wiki/Index-Block-Format#individual-index-block  I'll let you contemplate/research a solution but ping me if you get stuck.